File: crates/hologram-ffi/interfaces/python/examples/tensor_operations.py
#!/usr/bin/env python3
"""
Tensor Operations Example

This example demonstrates comprehensive tensor operations including:
- Tensor creation from buffers
- Tensor introspection (shape, strides, properties)
- Tensor operations (reshape, transpose, permute, view)
- Tensor slicing (select, narrow, slice)
- Matrix multiplication
- Broadcasting operations
- Proper cleanup
"""

import hologram_ffi as hg
import json

def main():
    print("=" * 60)
    print("Hologram FFI - Tensor Operations Example")
    print("=" * 60)
    
    # Create executor and buffers
    print("\n🏗️ Setting up Executor and Buffers:")
    try:
        executor_handle = hg.new_executor()
        print(f"✅ Created executor: {executor_handle}")
        
        # Create buffer for tensor
        buffer_handle = hg.executor_allocate_buffer(executor_handle, 24)  # 2x3x4 tensor
        print(f"✅ Created buffer: {buffer_handle}")
        
    except Exception as e:
        print(f"❌ Setup failed: {e}")
        return
    
    # Tensor creation
    print("\n📦 Tensor Creation:")
    try:
        # Create tensor from buffer with shape [2, 3, 4]
        shape = [2, 3, 4]
        tensor_handle = hg.tensor_from_buffer(buffer_handle, json.dumps(shape))
        print(f"✅ Created tensor with shape {shape}: {tensor_handle}")
        
        # Create tensor with custom strides
        strides = [12, 4, 1]  # Row-major strides
        buffer_handle2 = hg.executor_allocate_buffer(executor_handle, 24)
        tensor_handle2 = hg.tensor_from_buffer_with_strides(buffer_handle2, json.dumps(shape), json.dumps(strides))
        print(f"✅ Created tensor with custom strides {strides}: {tensor_handle2}")
        
    except Exception as e:
        print(f"❌ Tensor creation failed: {e}")
        return
    
    # Tensor introspection
    print("\n🔍 Tensor Introspection:")
    try:
        # Get tensor properties
        tensor_shape = hg.tensor_shape(tensor_handle)
        tensor_strides = hg.tensor_strides(tensor_handle)
        tensor_offset = hg.tensor_offset(tensor_handle)
        tensor_ndim = hg.tensor_ndim(tensor_handle)
        tensor_numel = hg.tensor_numel(tensor_handle)
        tensor_is_contiguous = hg.tensor_is_contiguous(tensor_handle)
        
        print(f"Tensor Properties:")
        print(f"  Shape: {json.loads(tensor_shape)}")
        print(f"  Strides: {json.loads(tensor_strides)}")
        print(f"  Offset: {tensor_offset}")
        print(f"  Dimensions: {tensor_ndim}")
        print(f"  Number of elements: {tensor_numel}")
        print(f"  Is contiguous: {tensor_is_contiguous}")
        
        # Get underlying buffer
        tensor_buffer_handle = hg.tensor_buffer(tensor_handle)
        print(f"  Underlying buffer: {tensor_buffer_handle}")
        
    except Exception as e:
        print(f"❌ Tensor introspection failed: {e}")
    
    # Tensor operations
    print("\n⚙️ Tensor Operations:")
    try:
        # Reshape tensor
        new_shape = [6, 4]
        reshaped_tensor = hg.tensor_reshape(tensor_handle, json.dumps(new_shape))
        reshaped_shape = hg.tensor_shape(reshaped_tensor)
        print(f"✅ Reshaped tensor to {new_shape}: {json.loads(reshaped_shape)}")
        
        # Cleanup reshaped tensor
        hg.tensor_cleanup(reshaped_tensor)
        
        # Create new tensor for transpose operation
        buffer_handle3 = hg.executor_allocate_buffer(executor_handle, 24)
        tensor_handle3 = hg.tensor_from_buffer(buffer_handle3, json.dumps([6, 4]))
        
        # Transpose tensor (for 2D tensors)
        transposed_tensor = hg.tensor_transpose(tensor_handle3)
        transposed_shape = hg.tensor_shape(transposed_tensor)
        print(f"✅ Transposed tensor: {json.loads(transposed_shape)}")
        
        # Cleanup transposed tensor
        hg.tensor_cleanup(transposed_tensor)
        
        # Create new tensor for permute operation
        buffer_handle4 = hg.executor_allocate_buffer(executor_handle, 24)
        tensor_handle4 = hg.tensor_from_buffer(buffer_handle4, json.dumps([4, 6]))
        
        # Permute dimensions
        perm_dims = [1, 0]  # Swap dimensions
        permuted_tensor = hg.tensor_permute(tensor_handle4, json.dumps(perm_dims))
        permuted_shape = hg.tensor_shape(permuted_tensor)
        print(f"✅ Permuted tensor with dims {perm_dims}: {json.loads(permuted_shape)}")
        
        # Cleanup permuted tensor
        hg.tensor_cleanup(permuted_tensor)
        
        # Create new tensor for view_1d operation
        buffer_handle5 = hg.executor_allocate_buffer(executor_handle, 24)
        tensor_handle5 = hg.tensor_from_buffer(buffer_handle5, json.dumps([6, 4]))
        
        # Create 1D view
        view_1d = hg.tensor_view_1d(tensor_handle5)
        view_shape = hg.tensor_shape(view_1d)
        print(f"✅ Created 1D view: {json.loads(view_shape)}")
        
        # Cleanup view_1d tensor
        hg.tensor_cleanup(view_1d)
        
    except Exception as e:
        print(f"❌ Tensor operations failed: {e}")
    
    # Tensor slicing operations
    print("\n✂️ Tensor Slicing Operations:")
    try:
        # Create a 3D tensor for slicing
        buffer_handle3 = hg.executor_allocate_buffer(executor_handle, 24)
        tensor_3d = hg.tensor_from_buffer(buffer_handle3, json.dumps([2, 3, 4]))
        
        # Select operation (reduces dimensionality)
        selected_tensor = hg.tensor_select(tensor_3d, 0, 1)  # Select index 1 along dim 0
        selected_shape = hg.tensor_shape(selected_tensor)
        print(f"✅ Selected tensor: {json.loads(selected_shape)}")
        
        # Narrow operation
        narrowed_tensor = hg.tensor_narrow(tensor_3d, 1, 1, 2)  # Narrow dim 1 from index 1, length 2
        narrowed_shape = hg.tensor_shape(narrowed_tensor)
        print(f"✅ Narrowed tensor: {json.loads(narrowed_shape)}")
        
        # Slice operation
        sliced_tensor = hg.tensor_slice(tensor_3d, 2, 1, 3, 1)  # Slice dim 2 from 1 to 3 with step 1
        sliced_shape = hg.tensor_shape(sliced_tensor)
        print(f"✅ Sliced tensor: {json.loads(sliced_shape)}")
        
        # Cleanup slicing tensors
        hg.tensor_cleanup(selected_tensor)
        hg.tensor_cleanup(narrowed_tensor)
        hg.tensor_cleanup(sliced_tensor)
        hg.tensor_cleanup(tensor_3d)
        
    except Exception as e:
        print(f"❌ Tensor slicing failed: {e}")
    
    # Matrix multiplication
    print("\n🔢 Matrix Multiplication:")
    try:
        # Create two 2D tensors for matrix multiplication
        buffer_a = hg.executor_allocate_buffer(executor_handle, 6)  # 2x3 matrix
        buffer_b = hg.executor_allocate_buffer(executor_handle, 8)  # 2x4 matrix
        
        tensor_a = hg.tensor_from_buffer(buffer_a, json.dumps([2, 3]))
        tensor_b = hg.tensor_from_buffer(buffer_b, json.dumps([2, 4]))
        
        # Matrix multiplication
        result_tensor = hg.tensor_matmul(executor_handle, tensor_a, tensor_b)
        result_shape = hg.tensor_shape(result_tensor)
        print(f"✅ Matrix multiplication result shape: {json.loads(result_shape)}")
        
        # Cleanup matrix multiplication tensors
        hg.tensor_cleanup(tensor_a)
        hg.tensor_cleanup(tensor_b)
        hg.tensor_cleanup(result_tensor)
        
    except Exception as e:
        print(f"❌ Matrix multiplication failed: {e}")
    
    # Broadcasting operations
    print("\n📡 Broadcasting Operations:")
    try:
        # Create tensors for broadcasting
        buffer_c = hg.executor_allocate_buffer(executor_handle, 4)  # 2x2 matrix
        buffer_d = hg.executor_allocate_buffer(executor_handle, 2)  # 2x1 vector
        
        tensor_c = hg.tensor_from_buffer(buffer_c, json.dumps([2, 2]))
        tensor_d = hg.tensor_from_buffer(buffer_d, json.dumps([2, 1]))
        
        # Check broadcast compatibility
        is_compatible = hg.tensor_is_broadcast_compatible(tensor_c, tensor_d)
        print(f"✅ Broadcast compatibility: {is_compatible}")
        
        # Get broadcast shape
        broadcast_shape = hg.tensor_broadcast_shape(tensor_c, tensor_d)
        print(f"✅ Broadcast shape: {json.loads(broadcast_shape)}")
        
        # Cleanup broadcasting tensors
        hg.tensor_cleanup(tensor_c)
        hg.tensor_cleanup(tensor_d)
        
    except Exception as e:
        print(f"❌ Broadcasting operations failed: {e}")
    
    # Cleanup
    print("\n🧹 Cleanup:")
    try:
        hg.tensor_cleanup(tensor_handle)
        hg.tensor_cleanup(tensor_handle2)
        hg.executor_cleanup(executor_handle)
        print("✅ All resources cleaned up successfully")
    except Exception as e:
        print(f"❌ Cleanup failed: {e}")
    
    print("\n✅ Tensor operations example completed!")

if __name__ == "__main__":
    main()

File: crates/hologram-ffi/interfaces/typescript/examples/tensor_operations.ts
/**
 * Tensor Operations Example
 * 
 * This example demonstrates comprehensive tensor operations including creation,
 * manipulation, reshaping, broadcasting, and advanced tensor operations.
 */

import * as hg from '../src/index';

console.log('🚀 Hologram FFI - Tensor Operations Example');
console.log('==========================================');

// Create executor
console.log('\n🏗️ Creating Executor:');
const executorHandle = hg.newExecutor();
console.log(`Created executor with handle: ${executorHandle}`);

// Create buffers for tensors
console.log('\n💾 Creating Buffers for Tensors:');
const bufferSize = 1000;
const bufferHandle1 = hg.executorAllocateBuffer(executorHandle, bufferSize);
const bufferHandle2 = hg.executorAllocateBuffer(executorHandle, bufferSize);
console.log(`Created buffers: ${bufferHandle1}, ${bufferHandle2}`);

// Fill buffers with test data
hg.bufferFill(bufferHandle1, 1.0, bufferSize);
hg.bufferFill(bufferHandle2, 2.0, bufferSize);
console.log('Filled buffers with test data');

// Create tensors from buffers
console.log('\n📦 Tensor Creation:');
const shape1 = JSON.stringify([10, 10, 10]); // 3D tensor
const shape2 = JSON.stringify([100, 10]);     // 2D tensor

const tensorHandle1 = hg.tensorFromBuffer(bufferHandle1, shape1);
const tensorHandle2 = hg.tensorFromBuffer(bufferHandle2, shape2);

console.log(`Created tensor 1 with handle: ${tensorHandle1}`);
console.log(`Created tensor 2 with handle: ${tensorHandle2}`);

// Tensor introspection
console.log('\n🔍 Tensor Introspection:');
console.log('\nTensor 1 Properties:');
const shape1Result = hg.tensorShape(tensorHandle1);
const strides1Result = hg.tensorStrides(tensorHandle1);
const offset1 = hg.tensorOffset(tensorHandle1);
const ndim1 = hg.tensorNdim(tensorHandle1);
const numel1 = hg.tensorNumel(tensorHandle1);
const isContiguous1 = hg.tensorIsContiguous(tensorHandle1);

console.log(`  Shape: ${shape1Result}`);
console.log(`  Strides: ${strides1Result}`);
console.log(`  Offset: ${offset1}`);
console.log(`  Dimensions: ${ndim1}`);
console.log(`  Number of elements: ${numel1}`);
console.log(`  Is contiguous: ${isContiguous1}`);

console.log('\nTensor 2 Properties:');
const shape2Result = hg.tensorShape(tensorHandle2);
const strides2Result = hg.tensorStrides(tensorHandle2);
const offset2 = hg.tensorOffset(tensorHandle2);
const ndim2 = hg.tensorNdim(tensorHandle2);
const numel2 = hg.tensorNumel(tensorHandle2);
const isContiguous2 = hg.tensorIsContiguous(tensorHandle2);

console.log(`  Shape: ${shape2Result}`);
console.log(`  Strides: ${strides2Result}`);
console.log(`  Offset: ${offset2}`);
console.log(`  Dimensions: ${ndim2}`);
console.log(`  Number of elements: ${numel2}`);
console.log(`  Is contiguous: ${isContiguous2}`);

// Tensor operations
console.log('\n🔄 Tensor Operations:');

// Reshape tensor
console.log('\n1. Reshaping:');
const newShape = JSON.stringify([50, 20]);
const reshapedTensor = hg.tensorReshape(tensorHandle1, newShape);
console.log(`Reshaped tensor handle: ${reshapedTensor}`);

const reshapedShape = hg.tensorShape(reshapedTensor);
console.log(`New shape: ${reshapedShape}`);

// Transpose tensor
console.log('\n2. Transposing:');
const transposedTensor = hg.tensorTranspose(tensorHandle2);
console.log(`Transposed tensor handle: ${transposedTensor}`);

const transposedShape = hg.tensorShape(transposedTensor);
console.log(`Transposed shape: ${transposedShape}`);

// Permute tensor
console.log('\n3. Permuting:');
const permDims = JSON.stringify([1, 0]); // Swap dimensions
const permutedTensor = hg.tensorPermute(tensorHandle2, permDims);
console.log(`Permuted tensor handle: ${permutedTensor}`);

const permutedShape = hg.tensorShape(permutedTensor);
console.log(`Permuted shape: ${permutedShape}`);

// Create 1D view
console.log('\n4. 1D View:');
const view1dTensor = hg.tensorView1d(tensorHandle1);
console.log(`1D view tensor handle: ${view1dTensor}`);

const view1dShape = hg.tensorShape(view1dTensor);
console.log(`1D view shape: ${view1dShape}`);

// Broadcasting operations
console.log('\n5. Broadcasting:');
const isCompatible = hg.tensorIsBroadcastCompatible(tensorHandle1, tensorHandle2);
console.log(`Tensors are broadcast compatible: ${isCompatible}`);

if (isCompatible) {
  const broadcastShape = hg.tensorBroadcastShape(tensorHandle1, tensorHandle2);
  console.log(`Broadcast shape: ${broadcastShape}`);
}

// Matrix multiplication
console.log('\n6. Matrix Multiplication:');
try {
  const matmulTensor = hg.tensorMatmul(executorHandle, tensorHandle1, tensorHandle2);
  console.log(`Matrix multiplication result tensor handle: ${matmulTensor}`);
  
  const matmulShape = hg.tensorShape(matmulTensor);
  console.log(`Matrix multiplication result shape: ${matmulShape}`);
  
  // Clean up matmul result
  hg.tensorCleanup(matmulTensor);
} catch (error) {
  console.log(`Matrix multiplication failed: ${error}`);
}

// Tensor slicing operations
console.log('\n7. Tensor Slicing:');

// Select operation
const selectedTensor = hg.tensorSelect(tensorHandle1, 0, 5);
console.log(`Selected tensor handle: ${selectedTensor}`);
const selectedShape = hg.tensorShape(selectedTensor);
console.log(`Selected tensor shape: ${selectedShape}`);

// Narrow operation
const narrowedTensor = hg.tensorNarrow(tensorHandle1, 0, 2, 6);
console.log(`Narrowed tensor handle: ${narrowedTensor}`);
const narrowedShape = hg.tensorShape(narrowedTensor);
console.log(`Narrowed tensor shape: ${narrowedShape}`);

// Slice operation
const slicedTensor = hg.tensorSlice(tensorHandle1, 0, 1, 8, 2);
console.log(`Sliced tensor handle: ${slicedTensor}`);
const slicedShape = hg.tensorShape(slicedTensor);
console.log(`Sliced tensor shape: ${slicedShape}`);

// Tensor buffer access
console.log('\n8. Buffer Access:');
const tensorBufferHandle = hg.tensorBuffer(tensorHandle1);
console.log(`Tensor buffer handle: ${tensorBufferHandle}`);

const tensorBufferMutHandle = hg.tensorBufferMut(tensorHandle1);
console.log(`Tensor mutable buffer handle: ${tensorBufferMutHandle}`);

// Create tensor with custom strides
console.log('\n9. Custom Strides:');
const customBufferHandle = hg.executorAllocateBuffer(executorHandle, 200);
hg.bufferFill(customBufferHandle, 3.0, 200);

const customShape = JSON.stringify([10, 20]);
const customStrides = JSON.stringify([20, 1]); // Row-major
const customStridesTensor = hg.tensorFromBufferWithStrides(customBufferHandle, customShape, customStrides);

console.log(`Custom strides tensor handle: ${customStridesTensor}`);
const customShapeResult = hg.tensorShape(customStridesTensor);
const customStridesResult = hg.tensorStrides(customStridesTensor);
console.log(`Custom strides tensor shape: ${customShapeResult}`);
console.log(`Custom strides tensor strides: ${customStridesResult}`);

// Performance test
console.log('\n⚡ Performance Test:');
const perfIterations = 100;
console.log(`Running ${perfIterations} tensor operations...`);

const perfStartTime = Date.now();
let currentTensor = tensorHandle1;

for (let i = 0; i < perfIterations; i++) {
  // Alternate between reshape and transpose
  if (i % 2 === 0) {
    const testShape = JSON.stringify([25, 40]);
    currentTensor = hg.tensorReshape(currentTensor, testShape);
  } else {
    currentTensor = hg.tensorTranspose(currentTensor);
  }
}

const perfEndTime = Date.now();
const perfDuration = perfEndTime - perfStartTime;
const opsPerSecond = (perfIterations / perfDuration) * 1000;

console.log(`Performance results:`);
console.log(`  Duration: ${perfDuration}ms`);
console.log(`  Operations per second: ${opsPerSecond.toFixed(0)}`);
console.log(`  Time per operation: ${(perfDuration / perfIterations).toFixed(3)}ms`);

// Cleanup
console.log('\n🧹 Cleanup:');
hg.tensorCleanup(tensorHandle1);
hg.tensorCleanup(tensorHandle2);
hg.tensorCleanup(reshapedTensor);
hg.tensorCleanup(transposedTensor);
hg.tensorCleanup(permutedTensor);
hg.tensorCleanup(view1dTensor);
hg.tensorCleanup(selectedTensor);
hg.tensorCleanup(narrowedTensor);
hg.tensorCleanup(slicedTensor);
hg.tensorCleanup(customStridesTensor);

hg.bufferCleanup(bufferHandle1);
hg.bufferCleanup(bufferHandle2);
hg.bufferCleanup(customBufferHandle);

hg.executorCleanup(executorHandle);

console.log('All tensors, buffers, and executor cleaned up');

console.log('\n✅ Tensor operations example completed successfully!');
